File: server/app/routers/tags.py
from typing import Any
from app.config import conf_pathname

from fastapi import APIRouter, Depends, Query
from sqlmodel import Session

from ..dependencies import get_tag_or_404, upsert_tag, soft_delete_tag
from app.database import get_session
from app.models import tag
from app.schemas import TagBase

router = APIRouter()

@router.get(conf_pathname()+"/v1/tags", response_model=list[TagBase])
def list_tags(*, session: Session = Depends(get_session), active_only: bool = Query(True)) -> list[TagBase]:
    return get_tag_or_404(session = session, model = tag, IsActive = active_only)

@router.get(conf_pathname()+"/v1/tags/{tag_id}", response_model=TagBase)
def get_tag(tag_id: int, session: Session = Depends(get_session)) -> TagBase:
    return get_tag_or_404(session = session, model = tag, tag_id = tag_id)

@router.get(conf_pathname()+"/v1/tags/by-name/{tag_Name}", response_model=list[TagBase])
def get_tag(tag_Name: str, session: Session = Depends(get_session), active_only: bool = Query(True)) -> list[TagBase]:
    return get_tag_or_404(session = session, model = tag, tag_name = tag_Name, IsActive = active_only)

@router.get(conf_pathname()+"/v1/tags/by-context/{tag_Context}", response_model=list[TagBase])
def get_tag(tag_Context: str, session: Session = Depends(get_session), active_only: bool = Query(True)) -> list[TagBase]:
    return get_tag_or_404(session = session, model = tag, tag_context = tag_Context, IsActive = active_only)

@router.get(conf_pathname()+"/v1/tags/by-name-context/{tag_Name}/{tag_Context}", response_model=list[TagBase])
def get_tag(tag_Name: str, tag_Context: str, session: Session = Depends(get_session), active_only: bool = Query(True)) -> list[TagBase]:
    return get_tag_or_404(session = session, model = tag, tag_name = tag_Name, tag_context = tag_Context, IsActive = active_only)

@router.post(conf_pathname()+"/v1/tags", response_model=TagBase)
def create_or_update_tag(payload: TagBase, session: Session = Depends(get_session)) -> TagBase:
    return upsert_tag(session = session, model = tag, payload = payload)

@router.delete(conf_pathname()+"/v1/tags/{tag_id}", response_model=TagBase)
def delete_tag(tag_id: int, session: Session = Depends(get_session)) -> TagBase:
    return soft_delete_tag(session = session, model = tag, tag_id = tag_id)

File: client-web/src/api/tags.ts
import { API_BASE } from '../config';

export async function listTags(IsActve: boolean = true) {
  const res = await fetch(`${API_BASE}/tags?active_only=${IsActve}`);
  if (!res.ok) {
    if (res.status != 404) throw new Error(`Failed to list tags: ${res.status}`);
    return "()";
  }
  return res.json();
}

export async function getTag(id: number) {
  const res = await fetch(`${API_BASE}/tags/${id}`);
  if (!res.ok) throw new Error(`Failed to load tag: ${res.status}`);
  return res.json();
}

export async function getTagByName(tagName: string) {
  const res = await fetch(`${API_BASE}/tags/by-name/${tagName}`);
  if (!res.ok) throw new Error(`Failed to load tag: ${res.status}`);
  return res.json();
}

export async function getTagByContext(tagContext: string) {
  const res = await fetch(`${API_BASE}/tags/by-context/${tagContext}`);
  if (!res.ok) throw new Error(`Failed to load tag: ${res.status}`);
  return res.json();
}

export async function saveTag(payload: any) {
  const res = await fetch(`${API_BASE}/tags`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  });
  if (!res.ok) throw new Error(`Failed to save tag: ${res.status}`);
  return res.json();
}

export async function deleteTag(id: number) {
  const res = await fetch(`${API_BASE}/tags/${id}`, {
    method: 'DELETE',
  });
  if (!res.ok) throw new Error(`Failed to delete tag: ${res.status}`);
  return res.json();
}

// Linked JobSpecs

export async function getJobSpecsByTag(tagId: number) {
  const res = await fetch(`${API_BASE}/roles/lnk/jobspecs-tag/${tagId}`);
  if (!res.ok) throw new Error('Failed to load tag job specs');
  return res.json();
}

export async function deleteJobSpecsByTag(tagId: number) {
  const res = await fetch(`${API_BASE}/roles/lnk/jobspecs-tag/${tagId}`, {
    method: 'DELETE',
  });
  if (!res.ok) throw new Error(`Failed to delete job specs tag: ${res.status}`);
  return res.json();
}

